### ISSUE TITLE
FQDN written to /etc/hosts causes problems for clustering systems

### ISSUE THREAD

**Issue by ubuntu-server-builder** [canonical/cloud-init#2215]
This bug was originally filed in Launchpad as [LP: #871966](https://bugs.launchpad.net/cloud-init/+bug/871966)
<details>
<summary>Launchpad details</summary>
<pre>
affected_projects = ['ubuntu-release-notes', 'cloud-init (Ubuntu)', 'cloud-init (Ubuntu Precise)', 'cassandra (Juju Charms Collection)']
assignee = smoser
assignee_name = Scott Moser
date_closed = 2012-04-11T04:07:45.010543+00:00
date_created = 2011-10-10T19:35:34.612691+00:00
date_fix_committed = 2011-12-20T03:52:41.187412+00:00
date_fix_released = 2012-04-11T04:07:45.010543+00:00
id = 871966
importance = medium
is_complete = True
lp_url = https://bugs.launchpad.net/cloud-init/+bug/871966
milestone = None
owner = smoser
owner_name = Scott Moser
private = False
status = fix_released
submitter = clint-fewbar
submitter_name = Clint Byrum
tags = []
duplicates = []
</pre>
</details>

_Launchpad user **Clint Byrum(clint-fewbar)** wrote on 2011-10-10T19:35:34.612691+00:00_

*** Ubuntu 11.10 Release Note ***

Cloud instances and servers pre-seeded with cloud-init will have their FQDN written to /etc/hosts and pointed to the IP 127.0.1.1. This may cause issues for daemons which try to listen on their hostname, rather than 0.0.0.0, as they will now only be reachable locally, rather than on the network address that their FQDN resolves to.

***

By writing the FQDN to /etc/hosts as resolving to 127.0.1.1, systems like Cassandra have a much harder time determining their address to communicate to other cluster members.

While some might see communicating your IP to others as  a bug, being able to use gethostname() and then resolving it to get the actual IP address of one's machine is fairly important.

Its my understanding that in resolving bug #802637 , the Debian networking docs were used as a guide:

http://www.debian.org/doc/manuals/debian-reference/ch05.en.html
Point 5.1.2 specifically.

It does suggest that one needs an FQDN in /etc/hosts.

However cloud-init should only set the addresss if it cannot be determined.

cloud-init should first try gethostbyname() on the FQDN. If it resolves, *do not write FQDN to /etc/hosts*. This assures that if it has been configured to be resolvable by some method in nsswitch.conf such as DNS or NIS or etc., it will not be overidden by /etc/hosts.

related bugs:
 bug 890501: EC2 cloud-init overwrites 127.0.1.1 in /etc/hosts on every reboot

**Comment by ubuntu-server-builder** [comment]
_Launchpad user **Launchpad Janitor(janitor)** wrote on 2011-10-10T19:45:38.976389+00:00_

Status changed to 'Confirmed' because the bug affects multiple users.

**Comment by ubuntu-server-builder** [comment]
_Launchpad user **James Page(james-page)** wrote on 2011-10-11T09:02:53.399115+00:00_

As a work-around until this is resolved in one way or another in juju I'm using:

  dig +short `unit-get private-address` 

dig deals with being passed an IP address nicely and will resolve a hostname correctly to the internal IP address of the instance in ec2 and openstack environments.

**Comment by ubuntu-server-builder** [comment]
_Launchpad user **James Page(james-page)** wrote on 2011-10-11T11:48:21.390561+00:00_

A bit more testing reveals that this solution does not work so well in a local juju environment; adapting with a check to see if an IP address is returned from private-address before trying to resolve using dig

**Comment by ubuntu-server-builder** [comment]
_Launchpad user **Scott Moser(smoser)** wrote on 2011-10-11T16:57:27+00:00_

On Tue, 11 Oct 2011, Clint Byrum wrote:

> Excerpts from Scott Moser's message of Tue Oct 11 00:18:27 UTC 2011:
> > gethostbyname(hostname) is non-determinable.
> > there could be multiple responses.  and indeterminable order.
> >
>
> Agreed, which is precisely why I believe cloud-init must leave the FQDN
> out of /etc/hosts if there is any response to this call.

That doesn't make any sense.  Its non-determinable, its non-determinable
when cloud-init runs as well as later on.
When cloud-init ran to decide if there was a response to this call, it may
give a single value, or may give 4 values, or may give no response.  That
may change later on when some other utility was asking.

**Comment by ubuntu-server-builder** [comment]
_Launchpad user **Clint Byrum(clint-fewbar)** wrote on 2011-10-11T17:30:22+00:00_

Excerpts from Scott Moser's message of Tue Oct 11 16:57:27 UTC 2011:
> On Tue, 11 Oct 2011, Clint Byrum wrote:
> 
> > Excerpts from Scott Moser's message of Tue Oct 11 00:18:27 UTC 2011:
> > > gethostbyname(hostname) is non-determinable.
> > > there could be multiple responses.  and indeterminable order.
> > >
> >
> > Agreed, which is precisely why I believe cloud-init must leave the FQDN
> > out of /etc/hosts if there is any response to this call.
> 
> That doesn't make any sense.  Its non-determinable, its non-determinable
> when cloud-init runs as well as later on.
> When cloud-init ran to decide if there was a response to this call, it may
> give a single value, or may give 4 values, or may give no response.  That
> may change later on when some other utility was asking.
> 

The number and content of responses is irrelevant, only that the count
of positive responses is at least one.

The idea isn't to cover the case for all time, its to cover the case
where users have not asked to have fqdn "fixed", yet we need to make a
best effort attempt to make sure it is resolvable during bootup.

If it can be resolved through the normal means, what reason do we have
to override that with a value like 127.0.1.1?

**Comment by ubuntu-server-builder** [comment]
_Launchpad user **Clint Byrum(clint-fewbar)** wrote on 2011-10-11T19:33:03.965547+00:00_

After discussing with Scott Moser, its agreed that this may cause issues, but not necessarily that it is a "bug" as much as a change in behavior that needs documenting. Adding a ubuntu-release-notes task with suggested release note.

**Comment by ubuntu-server-builder** [comment]
_Launchpad user **Scott Moser(smoser)** wrote on 2011-12-20T03:50:35.060592+00:00_

This is fix-commited in cloud-init in revision 491 (http://bazaar.launchpad.net/~cloud-init-dev/cloud-init/trunk/revision/491).

See the commit message there for more information.

**Comment by ubuntu-server-builder** [comment]
_Launchpad user **Jason X.(jasxun)** wrote on 2012-01-05T18:12:56.450825+00:00_

Will the fix also be available for Oneiric?

**Comment by ubuntu-server-builder** [comment]
_Launchpad user **Scott Moser(smoser)** wrote on 2012-01-05T21:01:50.680147+00:00_

Jason,
  I would actually like to not pull the fix back to oneiric.  As doing so would break people using oneiric who were expecting the behavior that is present there now.  At very least we have to think seriously about it and come up with a list of what types of users would be affected. Do you have any thoughts?

**Comment by ubuntu-server-builder** [comment]
_Launchpad user **Eric Hammond(esh)** wrote on 2012-01-06T10:03:11.001956+00:00_

Though I don't like the current way Oneiric manages /etc/hosts (and submitted related bug #890501) I agree with Scott that it is how Oneiric works on EC2 and changes could cause existing installations to break.  In fact, I have automated system code that works around the "bug" which would break if the behavior were fixed.  Not horrible for me personally, but I don't know how many might be in a similar situation.
